Failed assertions lack helpful context after #297

I love the intent behind #297 but it appears to have caused a very significant regression in functionality. By way of example, here is the output of https://github.com/onsi/gomega/releases/tag/v1.4.2 and ae19f1b56d53acc4d6a698412bfd1e92876163f6, when using `Expect(err).NotTo(HaveOccurred())` as an assertion.

Actual output (using https://github.com/onsi/gomega/releases/tag/v1.4.2): 
```
--- FAIL: TestGenerator (2.36s)
    --- FAIL: TestGenerator/Generator (2.36s)
        --- FAIL: TestGenerator/Generator/manually_constructing_a_fake/inspecting_the_target/the_target_is_a_function/IsInterface()_is_false (0.53s)
            assertion.go:75: 
                Expected error:
                    <*errors.errorString | 0xc004d279e0>: {
                        s: "cannot find package with function: HandlerFunc",
                    }
                    cannot find package with function: HandlerFunc
                not to have occurred
        --- FAIL: TestGenerator/Generator/manually_constructing_a_fake/inspecting_the_target/the_target_is_a_function/IsFunction()_is_true (0.57s)
            assertion.go:75: 
                Expected error:
                    <*errors.errorString | 0xc001c6b9d0>: {
                        s: "cannot find package with function: HandlerFunc",
                    }
                    cannot find package with function: HandlerFunc
                not to have occurred
FAIL
```

Output prior to #297 (specifically using ae19f1b56d53acc4d6a698412bfd1e92876163f6): 
```
--- FAIL: TestGenerator (2.44s)
    --- FAIL: TestGenerator/Generator (2.44s)
        --- FAIL: TestGenerator/Generator/manually_constructing_a_fake/inspecting_the_target/the_target_is_a_function/IsInterface()_is_false (0.63s)
            testing_t_support.go:22: 
                	/Users/joe/go/pkg/mod/github.com/onsi/gomega@v1.4.2-0.20180831124310-ae19f1b56d53/internal/assertion/assertion.go:69 +0x1ed
                github.com/onsi/gomega/internal/assertion.(*Assertion).NotTo(0xc00419bac0, 0x14404c0, 0x179ab30, 0x0, 0x0, 0x0, 0x1390e01)
                	/Users/joe/go/pkg/mod/github.com/onsi/gomega@v1.4.2-0.20180831124310-ae19f1b56d53/internal/assertion/assertion.go:43 +0xaf
                github.com/maxbrunsfeld/counterfeiter/generator.testGenerator.func3.2.3.1()
                	/Users/joe/projects/counterfeiter/generator/generator_internals_test.go:122 +0x25e
                github.com/sclevine/spec.run(0xc00143c3e0, 0x3, 0x4)
                	/Users/joe/go/pkg/mod/github.com/sclevine/spec@v1.0.0/spec.go:182 +0x33
                github.com/sclevine/spec.Run.func3(0xc000a35c00, 0xc00005a840, 0x4, 0x6, 0xc0000146a0, 0x4, 0x4, 0x5b9e789d, 0x1, 0x1, ...)
                	/Users/joe/go/pkg/mod/github.com/sclevine/spec@v1.0.0/spec.go:174 +0x2cc
                github.com/sclevine/spec.node.run.func1(0xc000a35c00)
                	/Users/joe/go/pkg/mod/github.com/sclevine/spec@v1.0.0/parser.go:132 +0x51
                testing.tRunner(0xc000a35c00, 0xc00143c260)
                	/usr/local/Cellar/go/1.11/libexec/src/testing/testing.go:827 +0xbf
                created by testing.(*T).Run
                	/usr/local/Cellar/go/1.11/libexec/src/testing/testing.go:878 +0x353
                
                Expected error:
                    <*errors.errorString | 0xc002113010>: {
                        s: "cannot find package with function: HandlerFunc",
                    }
                    cannot find package with function: HandlerFunc
                not to have occurred
        --- FAIL: TestGenerator/Generator/manually_constructing_a_fake/inspecting_the_target/the_target_is_a_function/IsFunction()_is_true (0.55s)
            testing_t_support.go:22: 
                	/Users/joe/go/pkg/mod/github.com/onsi/gomega@v1.4.2-0.20180831124310-ae19f1b56d53/internal/assertion/assertion.go:69 +0x1ed
                github.com/onsi/gomega/internal/assertion.(*Assertion).NotTo(0xc0034ca2c0, 0x14404c0, 0x179ab30, 0x0, 0x0, 0x0, 0x1390e01)
                	/Users/joe/go/pkg/mod/github.com/onsi/gomega@v1.4.2-0.20180831124310-ae19f1b56d53/internal/assertion/assertion.go:43 +0xaf
                github.com/maxbrunsfeld/counterfeiter/generator.testGenerator.func3.2.3.1()
                	/Users/joe/projects/counterfeiter/generator/generator_internals_test.go:122 +0x25e
                github.com/sclevine/spec.run(0xc00436f720, 0x3, 0x4)
                	/Users/joe/go/pkg/mod/github.com/sclevine/spec@v1.0.0/spec.go:182 +0x33
                github.com/sclevine/spec.Run.func3(0xc001315e00, 0xc00005a8a0, 0x4, 0x6, 0xc0000146c0, 0x4, 0x4, 0x5b9e789d, 0x1, 0x1, ...)
                	/Users/joe/go/pkg/mod/github.com/sclevine/spec@v1.0.0/spec.go:174 +0x2cc
                github.com/sclevine/spec.node.run.func1(0xc001315e00)
                	/Users/joe/go/pkg/mod/github.com/sclevine/spec@v1.0.0/parser.go:132 +0x51
                testing.tRunner(0xc001315e00, 0xc00436f300)
                	/usr/local/Cellar/go/1.11/libexec/src/testing/testing.go:827 +0xbf
                created by testing.(*T).Run
                	/usr/local/Cellar/go/1.11/libexec/src/testing/testing.go:878 +0x353
                
                Expected error:
                    <*errors.errorString | 0xc00342b5a0>: {
                        s: "cannot find package with function: HandlerFunc",
                    }
                    cannot find package with function: HandlerFunc
                not to have occurred
FAIL
```

As you can see, the output from v1.4.2 is indeed concise, just way too concise! There is no stack trace that lets you get back to the line where the failed assertion is made. I *want* to see the `	/Users/joe/projects/counterfeiter/generator/generator_internals_test.go:122` portion of the output, as that is the line where I am making the failed assertion. Jumping to here is essential for context and debugging.

You will note *I am running tests in the context of go 1.11 module mode*.

```
$ go version
go version go1.11 darwin/amd64
$ go env
GOARCH="amd64"
GOBIN=""
GOCACHE="/Users/joe/Library/Caches/go-build"
GOEXE=""
GOFLAGS=""
GOHOSTARCH="amd64"
GOHOSTOS="darwin"
GOOS="darwin"
GOPATH="/Users/joe/go"
GOPROXY=""
GORACE=""
GOROOT="/usr/local/Cellar/go/1.11/libexec"
GOTMPDIR=""
GOTOOLDIR="/usr/local/Cellar/go/1.11/libexec/pkg/tool/darwin_amd64"
GCCGO="gccgo"
CC="clang"
CXX="clang++"
CGO_ENABLED="1"
GOMOD="/Users/joe/projects/counterfeiter/go.mod"
CGO_CFLAGS="-g -O2"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-g -O2"
CGO_FFLAGS="-g -O2"
CGO_LDFLAGS="-g -O2"
PKG_CONFIG="pkg-config"
GOGCCFLAGS="-fPIC -m64 -pthread -fno-caret-diagnostics -Qunused-arguments -fmessage-length=0 -fdebug-prefix-map=/var/folders/zr/t34fdtsn6f5b5kz0smglx1dw0000gn/T/go-build245323484=/tmp/go-build -gno-record-gcc-switches -fno-common"
```

--- a/gomega_dsl.go
+++ b/gomega_dsl.go
@@ -46,9 +46,22 @@ func RegisterFailHandler(handler types.GomegaFailHandler) {
 		globalFailWrapper = nil
 		return
 	}
+
 	globalFailWrapper = &types.GomegaFailWrapper{
 		Fail:        handler,
 		TWithHelper: testingtsupport.EmptyTWithHelper{},
+	}
+}
+
+func RegisterFailHandlerWithT(t types.TWithHelper, handler types.GomegaFailHandler) {
+	if handler == nil {
+		globalFailWrapper = nil
+		return
+	}
+
+	globalFailWrapper = &types.GomegaFailWrapper{
+		Fail:        handler,
+		TWithHelper: t,
 	}
 }
 
@@ -74,7 +87,12 @@ func RegisterFailHandler(handler types.GomegaFailHandler) {
 //
 // (As an aside: Ginkgo gets around this limitation by running parallel tests in different *processes*).
 func RegisterTestingT(t types.GomegaTestingT) {
-	RegisterFailHandler(testingtsupport.BuildTestingTGomegaFailWrapper(t).Fail)
+	tWithHelper, hasHelper := t.(types.TWithHelper)
+	if !hasHelper {
+		RegisterFailHandler(testingtsupport.BuildTestingTGomegaFailWrapper(t).Fail)
+		return
+	}
+	RegisterFailHandlerWithT(tWithHelper, testingtsupport.BuildTestingTGomegaFailWrapper(t).Fail)
 }
 
 //InterceptGomegaHandlers runs a given callback and returns an array of

